lib/Conversion/ImportVerilog/Structure.cpp

```diff
@@ -470,11 +480,64 @@ ModuleLowering *
 Context::convertModuleHeader(const slang::ast::InstanceBodySymbol *module) {
   using slang::ast::ArgumentDirection;
   using slang::ast::MultiPortSymbol;
+  using slang::ast::ParameterSymbol;
   using slang::ast::PortSymbol;
+  using slang::ast::TypeParameterSymbol;
+
+  auto parameters = module->parameters;
+  bool hasModuleSame = false;
+  // If there is already exist a module that has the same name with this
+  // module ,has the same parent scope and has the same parameters we can
+  // define this module is a duplicate module
+  for (auto const &existingModule : modules) {
+    if (module->getDeclaringDefinition() ==
+        existingModule.getFirst()->getDeclaringDefinition()) {
+      auto moduleParameters = existingModule.getFirst()->parameters;
+      hasModuleSame = true;
+      for (auto it1 = parameters.begin(), it2 = moduleParameters.begin();
+           it1 != parameters.end() && it2 != moduleParameters.end();
+           it1++, it2++) {
+        // Parameters size different
+        if (it1 == parameters.end() || it2 == moduleParameters.end()) {
+          hasModuleSame = false;
+          break;
+        }
+        const auto *para1 = (*it1)->symbol.as_if<ParameterSymbol>();
+        const auto *para2 = (*it2)->symbol.as_if<ParameterSymbol>();
+        // Parameters kind different
+        if ((para1 == nullptr) ^ (para2 == nullptr)) {
+          hasModuleSame = false;
+          break;
+        }
+        // Compare ParameterSymbol
+        if (para1 != nullptr) {
+          hasModuleSame = para1->getValue() == para2->getValue();
+        }
+        // Compare TypeParameterSymbol
+        if (para1 == nullptr) {
+          auto para1Type = convertType(
+              (*it1)->symbol.as<TypeParameterSymbol>().getTypeAlias());
+          auto para2Type = convertType(
+              (*it2)->symbol.as<TypeParameterSymbol>().getTypeAlias());
+          hasModuleSame = para1Type == para2Type;
+        }
+        if (!hasModuleSame)
+          break;
+      }
+      if (hasModuleSame) {
+        module = existingModule.first;
+        break;
+      }
+    }
+  }
 
   auto &slot = modules[module];
-  if (slot)
+  if (slot) {
+    for (const auto &port : slot->ports)
+      slot->portsBySyntaxNode.insert({port.ast.getSyntax(), &port.ast});
```

Comment:
You don't have to do this every time a module is looked up. Instead, you can move this all the way to the bottom, just before `return &lowering;`. That way the `portsBySyntaxNode` is populated once when the lowering is created, and it remains valid. Then this line of code here can just be `if (slot) return slot.get();` again :slightly_smiling_face:

Reply:
Sorry, I was still thinking mapping  port symbol to symbol which should be done every time a module is looked up. You are very right, this map could be generated just when new lowering is generated. This will imporve running a lot.